### ISSUE TITLE
20230718 - day 29

### ISSUE THREAD

**Issue by tjkpolisher** [tjkpolisher/Bootcamp_TIL#29]
# 좋았던 점
* 대학원에 있을 때 처음 이 콜백들을 독학할 때 정말 힘들게 찾아가서 적용을 해도 정작 친절하게 주의사항 같은 걸 알려주는 곳이 없어서 이후로도 고생을 많이 했습니다. 진작에 이런 것도 좀 같이 알려주면 얼마나 좋아.... ~~내다버린 1년...~~

# 아쉬웠던 점
* 그래도 아직까지는 지나치게 이론에 편중된 감이 없지 않아 있네요. 사실 오늘 배운 성능개선 기법들은 대부분 이론보다는 실제 사용법, 특히 주의사항을 짚어주는 것이 더 도움이 될 것 같다는 느낌입니다.

# 배운 점
* 딥러닝 성능개선 기법'들': 실제로 딥러닝 모델을 만들다 보면, 절대적인 예측 성능은 물론이거니와 오버피팅을 억제하는 것이 지상 과제 중 하나로 꼽힙니다. 수학자와 컴퓨터공학자 선배님들의 오랜 연구 끝에 실무에서도 성능이 검증된 몇 가지 기법들이 소개되었습니다.
    * DNN 모델의 크기 변경: 가장 쉬운 방법 중 하나로, 은닉층의 개수를 늘리거나 각 레이어 별 유닛의 개수를 조절하는 방법을 사용합니다. 일반적으로 오버피팅일 경우 모델을 간단하게 만들고 언더피팅일 경우 모델을 크게 만드는 방식으로 성능을 튜닝합니다.
    * 드롭아웃(dropout): 드롭아웃은 학습 과정 중 매 에포크마다 일정 비율 만큼의 유닛을 **무작위로** 누락시킨 후 살아남아있는 유닛으로만 학습을 진행하는 방식입니다. 케라스의 레이어에 `dropout`이라는 이름으로 따로 구현되어 있으며, 여기서 `dropout_rate`라는 인수를 넣어서 비율을 조정합니다. 일반적으로 이 숫자가 0.5를 넘어가면 학습 성능이 극도로 저하되므로 최대 0.5 정도로만 입력합니다.</br>오버피팅의 원인 중 하나인 co-adaptiation을 줄이는 효과가 있지만, 매 에포크마다 일정 수의 유닛들을 무작위로 누락시킨 후 학습을 진행하기 때문에, 그만큼 학습 에포크를 더 많이 확보해야 충분한 예측 성능을 보인다는 점에 주의해야 합니다. 또한, 유닛의 누락은 오직 학습 과정 중에만 일어나며 검증(validation) 및 시험(test) 과정 중에는 뉴런의 누락을 일으키지 않습니다.
    * 배치 정규화(batch normalization): 각 레이어에서 출력된 값을 평균 0, 표준편차 1의 표준정규분포를 따르게끔 정규화함으로써 각 레이어의 입력분포를 균일하게 만들어 주는 기법입니다. 각 층을 통과할 때마다 입력 데이터 분포가 달라지는 현상인 '내부 공변량 변화'(internal covariance shift)를 막음으로써 성능 향상을 꾀하며, 드롭아웃과 마찬가지로 `BatchNormalization`이라는 API로 구현되어 있습니다.
    * Optimizer의 학습률 조정: 옵티마이저에 입력된 학습률 `learning_rate`의 값을 학습 중에 내부적으로 변화시킴으로써 성능을 향상시키는 방법입니다. 케라스의 콜백(callback) 함수들을 이용해 특정 조건(예를 들어 validation loss)이 만족될 경우(혹은 만족되지 못할 경우) 학습률을 일정 비율만큼 낮추는 방식을 일반적으로 사용합니다.
> TMI: 배치 정규화가 처음 발표된 논문에서 내부 공변량 변화를 막음으로써 성능 향상을 도모했다고 밝혔지만, 이후에 발표된 몇몇 후속 논문에 따르면 정작 의도했던 내부 공변량 변화가 아니라 알 수 없는 다른 요인을 억제함으로써 성능이 향상되었다고 밝혔다고도 합니다. ~~얻어 걸림??~~

# 앞으로 바라는 점
* 내일부터 CNN으로 들어갑니다. 제가 연구하면서 가장 피를 많이 본 부분이고, 연구 외적으로 진척이 되지 않은 탓에 연구하는 내내 주변 사람들로부터 상처를 많이 받았던 주제입니다. 이번 기회에 정말 제대로 배워 보고 싶습니다.
